### ISSUE TITLE
Triggering button on mouse event

### ISSUE THREAD

**Issue by hanek23** [mabe02/lanterna#452]
Hi,

When I was trying to enable mouse support in terminal I stumble upon one issue. I can focus all components in my application including buttons but I was not able to trigger them. Then I looked through code and found possibly a simple fix. There is a method in Button (handleKeyStroke) which determines if you have pressed "enter" and if so it triggers all listeners for that button. MouseEvent just goes "unnoticed" to the abstract parent. 

I suggest adding two conditions to that method. One being "OR keystroke is MouseEvent" other one being "AND button.isFocused()". That would cause first mouse event to pass through since the button is not yet focused (that is what the abstract parent will do) but the second one would actually trigger the button, which is what I believe is desirable.

Having said that I know I can just override that function and do it myself but I believe that should be default behavior. But since I do not see the whole picture of this library I decided to talk to you guys first.

Regards,
Petr

**Comment by hanek23** [comment]
Just realized that CheckBox also has its own "handleKeyStroke" and it can probably use this improvement as well.

**Comment by avl42** [comment]
I also think, the way forward is to make standard components deal with mouse-events in a sensible way.

There are just a couple of components, and for some of them (e.g. TextBox) mouse-handling is quite non-trivial.

If you could create a PR that would make most (if not all) of lanterna's components mouse-aware, I think that would be quite a welcome contribution.

E.g. TextBox'es could set the cursor to nearest character, and of course Buttons would trigger.

I like your approach that first click would just focus and only second click would do specific action, but it might be made "customizable" by setting a flag and depending on that flag either limit action to "already focused"-case or "do right-away, regardless of focus" or "set focus and do right-away".

**Comment by hanek23** [comment]
Yeah, I have already ran into a problem with ActionListBox which holds list of Runnables and did not see any, as you said trivial, way of handling that. But I will look into it more and I'll see what I can do.

**Comment by mabe02** [comment]
Yeah, I've kind of left all the mouse event handling out so far, because if done to one of the standard components we should really do it for all. But yes, adding mouse handling would be nice.

**Comment by hanek23** [comment]
Have you guys ever tried to run unix terminal inside of IDE (I am using Eclipse)? Not that I don't enjoy invoking maven install on lanterna and then assembly plugin on my project so I can test that one line that I have changed but it is starting to get little tedious, and since I am going thru code trying to understand everything around me I could really use debugging inside Eclipse. I have tried tell Eclipse to use java instead of javaw,  changing defaultTerminalFactory to give me unixTerminal everytime (that fails on privileges that I tried to work around but did not succeed either), running eclipse as root inside terminal and few other things. Or is there maybe any other way that I can test mouse support inside IDE?

**Comment by avl42** [comment]
There is indeed a solution to your problem:  The trick is the
TelnetTerminal.  you activate it by specifying
a port to the DefaultTerminalFactory, and maybe some more (you'll see it in
the sources).

Then (and I assume you're on Windows, as you mention java versus javaw)
you need a tool like "putty"
and that can also talk telnet.  You open a connection to localhost and the
selected port, and I'd expect it
to work...

I'm not entirely sure if the TelnetTerminal actually supports the
mouse-stuff, but I think it does.

**Comment by hanek23** [comment]
I am using linux, mint 19.1 to be precise, that is why I mentioned unix terminal. The java vs javaw is something I know has to be used on windows and it actually run in a swing terminal just like an IDE so I gave that a try, I do not know of any other way of launching that on Windows so I suppose mouse support for Window is, well out of the window. Anyway, thank you for your quick answer I will try to do that.

**Comment by hanek23** [comment]
It works like a charm. Thank you so much.

**Comment by hanek23** [comment]
Would it be possible for you guys to check out mouseSupport branch I have made before I make any pull request? I have couple of questions that I want to figure out before I make it and also I saw that my eclipse changed formating of the code so the commit diff is a mess. 
1. Mouse clicks seems to be +2 in both height and width of the components and I do not quite understand why, that maybe related to "magic number" 33 in MouseCharacterPattern where mouse clicks are parsed. What does that number mean?
2. What is the point of "test" class IssueXY if it is not really a automatic test? I have used it as kind of a showcase for all components where I test it myself but also created a Issue452Test class where I tried to test it automatically without launching TerminalScreen. I have been doing that in application that I am developing using lanterna and it works out pretty well so I would like to hear some thoughts on that matter as well.
3. avl42 mentioned some customizable flag that says if component (button) should be trigger at first click or second click (first click would just focus a button). Is that something I should work on when I am on it or should I leave that for another issue?

**Comment by hanek23** [comment]
Well I will my own question, at least the first one, if some poor soul did the same mistake as I did. I used toGlobal(TerminalPosition) method wrong. What I thought would happen is if I called it like toGlobal(getPosition()) inside component was to return its global coordinates of top left corner, which is obviously not the case. I added getGlobalPosition which does exactly that as a part of my pull request.
